plugins/out_s3/s3.c

```diff
@@ -1343,6 +1348,137 @@ static void cb_s3_upload(struct flb_config *config, void *data)
+    msgpack_object map;
+    msgpack_sbuffer tmp_sbuf;
+    msgpack_object key;
+    msgpack_object val;
+
+    /* Iterate the original buffer and perform adjustments */
+    records = flb_mp_count(data, bytes);
+    if (records <= 0) {
+        return NULL;
+    }
+
+    /* Create temporary msgpack buffer */
+    msgpack_sbuffer_init(&tmp_sbuf);
+
+    msgpack_unpacked_init(&result);
+    while (msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS) {
+        /* Each array must have two entries: time and record */
+        root = result.data;
+        if (root.type != MSGPACK_OBJECT_ARRAY) {
+            continue;
+        }
+        if (root.via.array.size != 2) {
+            continue;
+        }
+
+        /* Get the record/map */
+        map = root.via.array.ptr[1];
+        if (map.type != MSGPACK_OBJECT_MAP) {
+            continue;
+        }
+        map_size = map.via.map.size;
+
+        /* Extract log_key from record and append to output buffer */
+        for (i = 0; i < map_size; i++) {
+            key = map.via.map.ptr[i].key;
+            val = map.via.map.ptr[i].val;
+
+            if (key.type == MSGPACK_OBJECT_BIN) {
+                key_str  = (char *) key.via.bin.ptr;
+                key_str_size = key.via.bin.size;
+                check = FLB_TRUE;
+            }
+            if (key.type == MSGPACK_OBJECT_STR) {
+                key_str  = (char *) key.via.str.ptr;
+                key_str_size = key.via.str.size;
+                check = FLB_TRUE;
+            }
+
+            if (check == FLB_TRUE) {
+                if (strncmp(ctx->log_key, key_str, key_str_size) == 0) {
+                    found = FLB_TRUE;
+
+                    orig_val_buf = flb_malloc(bytes + bytes / 4);
+                    val_buf = orig_val_buf;
+                    if (val_buf == NULL) {
+                        flb_plg_error(ctx->ins, "Could not allocate enough "
+                                      "memory to read record");
+                        continue;
+                    }
+
+                    /* Fill val_buf with log_key contents */
+                    ret = flb_msgpack_to_json(val_buf, bytes + bytes / 4, &val);
+
+                    if (ret < 0) {
+                        flb_free(orig_val_buf);
+                        break;
+                    }
+
+                    /* Remove double quotes and add newline at end */
+                    if (val_buf[0] == '\"' && val_buf[strlen(val_buf)-1] == '\"') {
+                        val_buf[strlen(val_buf)-1] = '\0';
+                        val_buf++;
+                    }
+
+                    /* Create or concatenate to out_buf for full record chunk */
+                    if (out_buf == NULL) {
+                        out_buf = flb_sds_create(val_buf);
+                    } else {
+                        out_buf = flb_sds_cat(out_buf, val_buf, strlen(val_buf));
+                    }
+                    out_buf = flb_sds_cat(out_buf, "\n", 1);
```

Comment:
all of these functions can allocate memory IIRC, and need checks that they were successful